task/download/download.go

```diff
@@ -44,40 +45,66 @@ type downloader struct {
 // whether it should be re-cloned.
 func getHashOfRepo(repo *task.Repository) string {
 	data := fmt.Sprintf("%s|%s|%s|%s", repo.Clone, repo.Ref, repo.Sha, repo.Download)
+	return getHash(data)
+}
+
+func getHash(s string) string {
 	hash := sha256.New()
-	hash.Write([]byte(data))
+	hash.Write([]byte(s))
 	return hex.EncodeToString(hash.Sum(nil))
 }
 
-func (d *downloader) Download(ctx context.Context, repo *task.Repository) (string, error) {
-	if repo == nil {
-		return "", errors.New("no repository provided to download")
+func (d *downloader) Download(ctx context.Context, repo *task.Repository, urls *task.ExecutableUrls) (string, error) {
+	if urls != nil {
+		return d.handleDownloadExecutable(ctx, urls)
+	} else if repo != nil {
+		return d.handleDownloadRepo(ctx, repo)
 	}
-	log := logger.FromContext(ctx)
+	return "", errors.New("no repository or executable urls provided to download")
+}
+
+func (d *downloader) handleDownloadExecutable(ctx context.Context, urls *task.ExecutableUrls) (string, error) {
+	osWithArch := fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH)
+	url, ok := (*urls)[osWithArch]
+	if !ok {
+		return "", fmt.Errorf("os and architecture [%s] is not specified in task.yml file", osWithArch)
+	}
+
+	dest := filepath.Join(d.getBaseDownloadDir(), getHash(url))
+
+	if cacheHit := d.isCacheHit(ctx, dest); cacheHit {
+		// exit if the artifact destination already exists
+		return d.getDownloadPath(url, dest), nil
+	}
+
+	binpath, err := d.downloadFile(ctx, url, dest)
+	if err != nil {
+		// remove the destination directory if downloading fails so it can be retried
+		os.RemoveAll(dest)
+		return "", err
+	}
+
+	err = os.Chmod(binpath, 0777)
+	if err != nil {
+		return "", fmt.Errorf("failed to set executable flag in task file [%s]: %w", binpath, err)
+	}
+	return binpath, nil
+}
 
+func (d *downloader) handleDownloadRepo(ctx context.Context, repo *task.Repository) (string, error) {
```

Comment:
I did some refactoring here, to avoid code repetition when dealing with downloading an executable file. But the logic for repo downloading is unchanged.